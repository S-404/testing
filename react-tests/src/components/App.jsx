import { useEffect, useState } from 'react'


function App() {
    const [count, setCount] = useState(0)
    const [data, setData] = useState(null)
    const [toggle, setToggle] = useState(false)
    const [inputVal, setInputVal] = useState('')

    const onClick = () => setToggle(prevState => !prevState)

    useEffect(() => {
        setTimeout(() => {
            setData({})
        }, 100)
    })

    return (
        <>
            <div>
                <a href="https://vitejs.dev" target="_blank">
                    hello
                </a>
            </div>
            <h1>Vite + React</h1>

            <div className="card">
                <button data-testid="increment-button" onClick={() => setCount((count) => count + 1)}>
                    count is {count}
                </button>
                <p>
                    Edit <code>src/App.jsx</code> and save to test HMR
                </p>
            </div>
            <p className="read-the-docs">
                Click on the Vite and React logos to learn more
            </p>

            {data && <div>has data</div>}
            {toggle && <div data-testid="toggle-element">toggle</div>}
            <button data-testid="toggle-button" onClick={onClick}>
                toggle-click
            </button>

            <div data-testid="input-value">{inputVal}</div>
            <input
                data-testid="input"
                value={inputVal}
                onChange={(e) => setInputVal(e.target.value)}
            />
        </>
    )
}

export default App
