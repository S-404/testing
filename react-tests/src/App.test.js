import { fireEvent, render, screen } from '@testing-library/react'
import App from './App.jsx'
import '@testing-library/jest-dom'


describe('app test', () => {
    test('renders hello', () => {
        render(<App/>)
        const element = screen.getByText(/hello/i)
        expect(element).toBeInTheDocument()
    })

    test('renders button', () => {
        render(<App/>)
        const btn = screen.getByRole('button')
        expect(btn).toHaveTextContent('count is')
        expect(btn).toMatchSnapshot()
        screen.debug()
    })

    test('query: no element', () => {
        render(<App/>)
        const element = screen.queryByText(/hello2/i)
        expect(element).toBeNull()
    })

    test('has async data', async () => {
        render(<App/>)
        screen.debug()
        const element = await screen.findByText(/has data/i)
        expect(element).toBeInTheDocument()
        screen.debug()
    })

    test('toggle onClick', async () => {
        render(<App/>)
        const toggleButton = screen.getByTestId('toggle-button')

        expect(screen.queryByTestId('toggle-element')).toBeNull()
        fireEvent.click(toggleButton)
        expect(screen.queryByTestId('toggle-element')).toBeInTheDocument()
        fireEvent.click(toggleButton)
        expect(screen.queryByTestId('toggle-element')).toBeNull()
    })

    test('input text', async () => {
        render(<App/>)
        const input = screen.getByTestId('input')

        expect(screen.getByTestId('input-value')).toContainHTML('')
        fireEvent.input(input, {
            target: {value: '123abc'},
        })
        expect(screen.getByTestId('input-value')).toContainHTML('123abc')

    })
})

