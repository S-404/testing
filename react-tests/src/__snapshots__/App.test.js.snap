// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app test renders button 1`] = `
<button>
  count is 
  0
</button>
`;
